rfcs/007-bytecode-interpreter.md

```diff
@@ -0,0 +1,834 @@
+
+#### Bytecode
+
+The second representation is a compact and flat (array-like) representation.
+This code is composed of instructions, that will be interpreted by the virtual
+machine, heap-allocated values and additional metadata (span, etc.). The memory
+representation of those components is detailed in later sections.
+
+We will use a simple unsigned integer type to represent code addresses. To avoid
+the confusion with native pointers, we will simply call them indices instead of
+pointer in the rest of this document.
+
+### Memory representation of values
+
+#### Uniform representation
+
+Here is a list of Nickel values that we need to represent at runtime:
+
+- Null
+- Booleans
+- Number (54 bytes on 64bits arch)
+- String (24 bytes on 64bits arch)
+- Sealing Key
+- Foreign value (opaque)
+- Enum variant
+- Record
+- Array
+- Label
+- Custom contract
+- Function
+
+The only primitive values that are small enough to be inlined are `null` and
+`boolean`. Other values are boxed, represented as a pointer to a block. We can
+use the least significant bit of the pointer (pointer tagging) on any
+architecture where pointers are at least 2-bytes aligned (every architecture
+that matters for Nickel in practice) to inline `null` and booleans.
+
+In practice, the pointer is a reference-counted pointer (`Rc<_>` in Rust) that
+goes to a block of memory of variable size, with a 1-word header holding the
+discriminant (and possibly more information if needed), followed by the span and
+then the actual data. In total, accounting for `Rc`, there will be two words
+(the weak and the strong counter) before the header, so the global layout of a
+block is:
+
+```text
+| handled by std::rc::Rc  | actual block            |
+-----------------------------------------------------
+| weak (1w) | strong (1w) | head (1w) | span | data |
+-----------------------------------------------------
+```
+
+In a way, ignoring the inlined values, our main representation is one big enum
+of variants with exactly one argument that is pointer. In this specific case, we
+can use the proposed optimized representation where the tag is stored alongside
+the pointee, instead of alongside the pointer as with the standard Rust enum
+representation. This saves a lot of space at the cost of making the pointee
+dynamically-sized, which is less nice to handle in Rust but isn't a blocker
+either.
+
+#### Records and arrays
+
+We leave the precise representation of records and arrays open for now. Their
+actual representation doesn't impact the rest of the proposal and can be
+optimized at will in the future. The current representation works fine.
+
+One simple optimization worth mentioning would be to special case the empty
+record and the empty array, for example with `enum Record { Empty,
+NonEmpty(RecordData) }`. This should use the same space as `RecordData` in Rust
+(if `RecordData` is a pointer, at least) and save an allocation for empty
+structures.
```

Comment:
I'm a bit confused about what discriminants are present in the "top-level" representation (I'm not sure what's the right term, but I'm talking about the word-sized thing that packs in a pointer with some discriminant). Above, it sounded like we were only going to inline `null` and `boolean` in the discriminant; here, you're also proposing the put empty records and empty arrays? It seems like there isn't enough room in the pointer alignment for all of these.

By the way, x86_64, aarch64, and riscv-64 all max out at 48 bits of address space. So on these architectures we can pack lots more stuff at the most-significant end of the top-level representation.

Reply:
> By the way, x86_64, aarch64, and riscv-64 all max out at 48 bits of address space. So on these architectures we can pack lots more stuff at the most-significant end of the top-level representation.

Yes, but it's less portable. Although we don't need Nickel to run on embedded, I think we can do with one bit for now, and explore those other possibilities later.

> I'm a bit confused about what discriminants are present in the "top-level" representation (I'm not sure what's the right term, but I'm talking about the word-sized thing that packs in a pointer with some discriminant). Above, it sounded like we were only going to inline null and boolean in the discriminant; here, you're also proposing the put empty records and empty arrays? It seems like there isn't enough room in the pointer alignment for all of these.

Your first understanding is right. At the top-level, there is only one discriminant for `bool`, `null` and `pointer`. If we follow the pointer, we find many representations: arrays, numbers, etc. Here I'm talking about the representation of the pointee, which can itself be a pointer to something else (typically I guess your immutable vec representation would be mostly a pointer to the root plus some parameters).  Somehow the 1-word representation can hold any data, and I'm talking about this specific data when the pointee represents a record. Does that make sense?

Reply:
Also, as we need a discriminant at the beginning of the pointee (is it an Array? a Record? etc.) that will need to be aligned (although it might be merged with some other metadata as well), we'll probably have some more space here, and can even special case `EmptyArray` and `EmptyRecord` as special discriminants, instead of bothering making `Record` actually an enum.

Reply:
Even on 32-bit, we can have lots more values at the top-level, right? Assuming our pointers are all 4-byte aligned, anything ending in 01, 10, and 11 is not a pointer. But then for each of those non-pointer values we still have 30 bits left to store actual data. So couldn't we have variants for `EmptyArray` and `EmptyRecord` without even following that single top-level pointer? And we'd still have room for small integers...

Reply:
Sure, we have a lot more room. It's just that you can't encode anything that needs at least a full machine word or more - typically OCaml needs to have 31-bits and 63-bits integers so that they can unbox them, which is the case of most other non trivial data structures. But indeed special values like empty stuff could in theory be also directly put that top-level.

Reply:
Ok, makes sense. I was just confused about what discriminants were where. Our current `Vector` has a `root: Option<Rc<Node>>`, so it's already avoiding allocation for empty arrays.

Reply:
I've tried to make this part clearer: instead of mentioning empty records, I've moved the discussion to when we introduce the first level representation, and list many possible inlining (small numbers, small strings) that include empty records and arrays. It feels less ad-hoc.